File: api/routers/ai.py
"""
AI assistant router.

Endpoints:
  POST /ai/generate  — SSE stream of Jinja2 template tokens
  GET  /ai/status    — whether AI is configured (uses DB-resolved settings)

AI config is resolved with DB-wins-over-env precedence via settings_service.
"""

from __future__ import annotations

import json

from fastapi import APIRouter, Depends, HTTPException, status
from fastapi.responses import StreamingResponse
from sqlalchemy.ext.asyncio import AsyncSession

from api.core.auth import TokenData, get_current_user
from api.database import get_db
from api.schemas.ai import AIGenerateRequest
from api.services.ai_service import build_system_prompt, get_provider_from_config
from api.services.settings_service import get_resolved_ai_config

router = APIRouter()


@router.post(
    "/generate",
    summary="Stream AI-generated Jinja2 template body",
    description=(
        "Accepts a natural-language prompt and streams back a Jinja2 template body "
        "as Server-Sent Events. Each SSE `data:` line carries a JSON-encoded text chunk. "
        "Returns `data: [DONE]` when the stream is complete. "
        "Returns HTTP 503 when AI is not configured."
    ),
)
async def generate(
    body: AIGenerateRequest,
    current_user: TokenData = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> StreamingResponse:
    cfg = await get_resolved_ai_config(db, current_user.org_id)
    try:
        provider = get_provider_from_config(**cfg)
    except ValueError as exc:
        raise HTTPException(status_code=status.HTTP_503_SERVICE_UNAVAILABLE, detail=str(exc))

    if provider is None:
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail="AI assistant is not configured. Set AI_PROVIDER in System → Settings or in your environment.",
        )

    system = build_system_prompt(
        registered_params=body.registered_params,
        custom_filters=body.custom_filters,
        existing_body=body.existing_body,
    )

    async def event_stream():
        try:
            async for chunk in provider.stream(system, body.prompt):
                yield f"data: {json.dumps(chunk)}\n\n"
        except Exception as exc:  # noqa: BLE001
            yield f"data: {json.dumps({'__error__': str(exc)})}\n\n"
        finally:
            yield "data: [DONE]\n\n"

    return StreamingResponse(
        event_stream(),
        media_type="text/event-stream",
        headers={
            "Cache-Control": "no-cache",
            "X-Accel-Buffering": "no",
        },
    )


@router.get(
    "/status",
    summary="Check whether the AI assistant is configured",
)
async def ai_status(
    current_user: TokenData = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> dict:
    cfg = await get_resolved_ai_config(db, current_user.org_id)
    try:
        provider = get_provider_from_config(**cfg)
        enabled = provider is not None
        error = None
    except ValueError as exc:
        enabled = False
        error = str(exc)

    return {
        "enabled": enabled,
        "provider": cfg["provider"] or None,
        "model": cfg["model"] if enabled else None,
        "error": error,
    }

File: frontend/src/api/ai.ts
import { getToken } from './client'

const API_BASE = (import.meta.env.VITE_API_URL ?? '/api') as string

export interface AIGenerateRequest {
  prompt: string
  registered_params?: string[]
  custom_filters?: string[]
  existing_body?: string
}

export interface AIStatus {
  enabled: boolean
  provider: string | null
  model: string | null
  error: string | null
}

/** Check whether the AI assistant is configured on the server. */
export async function getAIStatus(): Promise<AIStatus> {
  const token = getToken()
  const res = await fetch(`${API_BASE}/ai/status`, {
    headers: token ? { Authorization: `Bearer ${token}` } : {},
  })
  if (!res.ok) throw new Error('Failed to fetch AI status')
  return res.json()
}

/**
 * Stream Jinja2 template body tokens from the AI assistant.
 *
 * Yields plain text chunks as they arrive. Each `data:` SSE line carries
 * a JSON-encoded string. The stream ends when `data: [DONE]` is received.
 *
 * Throws if the server returns an error status or streams an `__error__` event.
 */
export async function* streamAIGenerate(
  request: AIGenerateRequest,
  signal?: AbortSignal,
): AsyncGenerator<string> {
  const token = getToken()
  const response = await fetch(`${API_BASE}/ai/generate`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      ...(token ? { Authorization: `Bearer ${token}` } : {}),
    },
    body: JSON.stringify(request),
    signal,
  })

  if (!response.ok) {
    const err = await response.json().catch(() => ({ detail: 'AI generation failed' }))
    throw new Error(err.detail ?? 'AI generation failed')
  }

  const reader = response.body!.getReader()
  const decoder = new TextDecoder()
  let buffer = ''

  while (true) {
    const { done, value } = await reader.read()
    if (done) break

    buffer += decoder.decode(value, { stream: true })
    const lines = buffer.split('\n')
    buffer = lines.pop() ?? ''

    for (const line of lines) {
      if (!line.startsWith('data:')) continue
      const data = line.slice(5).trim()
      if (!data || data === '[DONE]') continue

      let parsed: unknown
      try {
        parsed = JSON.parse(data)
      } catch {
        continue
      }

      // Server-side error surfaced as a special event
      if (typeof parsed === 'object' && parsed !== null && '__error__' in parsed) {
        throw new Error((parsed as { __error__: string }).__error__)
      }

      if (typeof parsed === 'string') {
        yield parsed
      }
    }
  }
}
